napari/layers/utils/layer_utils.py

```diff
@@ -204,81 +216,473 @@ def _nanmax(array):
+        The y chunk start indices in array indices.
+    x_start_indices: list[int]
+        The x chunk start indices in array indices.
+    """
+    if chunk_shape:
+        plane_size = chunk_shape[:-offset]
+        # Go from chunk indices to pixel indices.
+        plane_indices = [
+            (plane_index[size_index] * size,)
+            for size_index, size in enumerate(plane_size)
+            for plane_index in plane_indices
+        ]
+
+        y_start_indices = _get_start_indices(plane_shape[0], chunk_size_y)
+        x_start_indices = _get_start_indices(plane_shape[1], chunk_size_x)
+    else:
+        y_start_indices = _get_start_indices(plane_shape[0])
+        x_start_indices = _get_start_indices(plane_shape[1])
+
+    return plane_indices, y_start_indices, x_start_indices
+
+
+def _get_crop_slices(
+    shape: Sequence[int],
+    plane_indices: Sequence[Sequence[int]],
+    offset: int,
+    chunk_shape: None | Tuple[int, ...] = None,
+) -> List[Tuple[Union[int, slice], ...]]:
+    """
+    Get the crop slices to be used for determining contrast limits when data is larger than the pixel threshold.
+
+    Currently, we support a maximum of 9 crops per plane taken for both the y and x dimension as the first quantile,
+    second quantile and the third quantile. In case of having a numpy array this number is never changed, only the size
+    of the crops changes. In case of having chunked data, this number can be changed dependent on the PIXEL_THRESHOLD
+    and the chunk size. The overall behaviour if with 9 chunks we go over the PIXEL_THRESHOLD is that we first sacrifice
+    in the number of chunks per plane and only when this still causes to go over the PIXEL_TRESHOLD do we sacrifice on
+    also sampling planes. If we can support n fold 9 chunks we will extend the crops by n.
+
+    Parameters
+    ----------
+    shape: Sequence[int]
+        Either the shape of the raw data (in pixels) or the block shape of the data (chunks)
+    plane_indices: Sequence[Sequence[int]]
+        Bottom, middle and top plane or single plane index for each non-visible dimension.
+    offset: int
+        Number of visible dimensions.
+    chunk_shape: None | Tuple[int]
+        The size per dimension of the chunks.
+
+    Returns
+    -------
+    slices: Union[None, List[Tuple[Union[int, slice], ...]]]
+        A list of crop slices.
+    """
+    plane_shape = shape[-offset:]
+
+    (
+        max_chunks_per_plane,
+        max_allowed_chunks,
+        chunk_size_y,
+        chunk_size_x,
+    ) = _calculate_chunk_parameters(plane_indices, offset, chunk_shape)
+
+    plane_indices, y_start_indices, x_start_indices = _get_pixel_start_indices(
+        plane_indices,
+        plane_shape,
+        offset,
+        chunk_size_y,
+        chunk_size_x,
+        chunk_shape,
+    )
+
+    start_indices = [
+        (y_start, x_start)
+        for y_start in y_start_indices
+        for x_start in x_start_indices
+    ]
```

Comment:
`itertools.product`?

Reply:
I would be fine either way, but does it really improve speed in this case and is it worth it with regards to readability? I don't know I am neutral here.

Reply:
switched to product with comment behind that explains it.